url8d1df200a8_gilleain_signatures_tgz-pJ8-signature_chemistry_LargeMoleculeTestJ8/src/signature/DAG.java

```diff
@@ -0,0 +1,682 @@
+			this.vertexIndex = vertexIndex;
+			this.layer = layer;
+			this.parents = new ArrayList<Node>();
+			this.children = new ArrayList<Node>();
+			this.edgeColors = new HashMap<Integer, Integer>();
+		}
+	
+        public void addParent(Node node) {
+			this.parents.add(node);
+		}
+		
+		public void addChild(Node node) {
+			this.children.add(node);
+		}
+		
+		public void addEdgeColor(int partnerIndex, int edgeColor) {
+            this.edgeColors.put(partnerIndex, edgeColor);
+        }
+		
+		public void accept(DAGVisitor visitor) {
+		    visitor.visit(this);
+		}
+		
+		/* (non-Javadoc)
+		 * @see java.lang.Object#toString()
+		 */
+		public String toString() {
+		    StringBuffer parentString = new StringBuffer();
+		    parentString.append('[');
+		    for (Node parent : this.parents) {
+		        parentString.append(parent.vertexIndex).append(',');
+		    }
+		    if (parentString.length() > 1) {
+		        parentString.setCharAt(parentString.length() - 1, ']');
+		    } else {
+		        parentString.append(']');
+		    }
+		    StringBuffer childString = new StringBuffer();
+		    childString.append('[');
+            for (Node child : this.children) {
+                childString.append(child.vertexIndex).append(',');
+            }
+            if (childString.length() > 1) {
+                childString.setCharAt(childString.length() - 1, ']');    
+            } else {
+                childString.append(']');
+            }
+            
+            return vertexIndex + " " 
+                  + " (" + parentString + ", " + childString + ")";
+		}
+
+	}
+	
+	/**
+	 * An arc of the directed acyclic graph.
+	 *
+	 */
+	public class Arc {
+		
+		public final int a;
+		
+		public final int b;
+		
+		public Arc(int a, int b) {
+			this.a = a;
+			this.b = b;
+		}
+		
+		public boolean equals(Object other) {
+			if (other instanceof Arc) {
+				Arc o = (Arc) other;
+				return (this.a == o.a && this.b == o.b)
+					|| (this.a == o.b && this.b == o.a);
+			} else {
+				return false;
+			}
+		}
+	}
+	
+	/**
+	 * Comparator for nodes based on String labels.
+	 *
+	 */
+	public class NodeStringLabelComparator implements Comparator<Node> {
+	    
+	    /**
+	     * The labels for vertices.
+	     */
+	    public String[] vertexLabels;
+	    
+	    public NodeStringLabelComparator(String[] vertexLabels) {
+	        this.vertexLabels = vertexLabels;
+	    }
+
+        public int compare(Node o1, Node o2) {
+            String o1s = this.vertexLabels[o1.vertexIndex];
+            String o2s = this.vertexLabels[o2.vertexIndex];
+            int c = o1s.compareTo(o2s);
+            if (c == 0) {
+                if (o1.invariant < o2.invariant) {
+                    return -1;
+                } else if (o1.invariant > o2.invariant) {
+                    return 1;
+                } else {
+                    return 0;
+                }
+            } else {
+                return c;
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 170:180, 203:213.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.